src/operator/nn/mkldnn/mkldnn_base.cc

```diff
@@ -270,9 +270,24 @@ void FallBackCompute(FCompute fn, const nnvm::NodeAttrs &attrs,
                      const std::vector<OpReqType> &req,
                      const std::vector<NDArray> &outputs) {
   std::vector<TBlob> in_blobs(inputs.size());
+  std::vector<NDArray> in_bufs;
   for (size_t i = 0; i < in_blobs.size(); i++) {
+    // If the input data isn't stored in the default format, we shouldn't
+    // call data() directly, which will change the layout of the NDArray.
+    // Instead, we should save the converted data in another NDArray.
+    // TODO(zhengda) we should use temp space to save the converted data.
+    if (inputs[i].IsDefaultData()) {
       in_blobs[i] = inputs[i].data();
+    } else {
+      in_bufs.emplace_back(inputs[i].shape(), inputs[i].ctx(),
```

Comment:
nit: it would be nice to call in_bufs.reserve(in_blobs.size()) on this first pass (ie if in_bufs.empty())
If there is often more than one in_bufs to be inserted

Reply:
I saw reserve() being used in many places in MXNet. I'm always wondering how much benefit it brings. Or is this a convention?

Reply:
It prevents reallocations, heap fragmentation, and contention if you know the size, it's definitely good to use reserve in advance.

Reply:
when you add items to a vector, when it runs out of space for the items (which generally happens on the second item sometimes), it allocates memory for 2x the current number of items and then copies all of the items (copy constructor or move, depending on a number of things), into to the new space and then frees the old space. this can happen over and over again, and the whole reallocate and copy can be very expensive. they tend to even show up as hotspots in vtune. so if you know how many items a vector may have, it’s always a good idea to call reserve() because otherwise you are guaranteed a lot of extra reallocing And copying.

Reply:
implementations vary, but this gives a general idea:

http://www.drdobbs.com/c-made-easier-how-vectors-grow/184401375